servers/services/src/main/java/org/projectnessie/services/impl/NamespaceApiImpl.java

```diff
@@ -0,0 +1,279 @@
+ * limitations under the License.
+ */
+package org.projectnessie.services.impl;
+
+import com.google.common.collect.Sets;
+import java.security.Principal;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.projectnessie.api.NamespaceApi;
+import org.projectnessie.api.params.NamespaceParams;
+import org.projectnessie.error.NessieNamespaceAlreadyExistsException;
+import org.projectnessie.error.NessieNamespaceNotEmptyException;
+import org.projectnessie.error.NessieNamespaceNotFoundException;
+import org.projectnessie.error.NessieReferenceNotFoundException;
+import org.projectnessie.model.CommitMeta;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.Content.Type;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.GetNamespacesResponse;
+import org.projectnessie.model.ImmutableGetNamespacesResponse;
+import org.projectnessie.model.Namespace;
+import org.projectnessie.model.Operation.Delete;
+import org.projectnessie.model.Operation.Put;
+import org.projectnessie.services.authz.Authorizer;
+import org.projectnessie.services.config.ServerConfig;
+import org.projectnessie.versioned.BranchName;
+import org.projectnessie.versioned.Key;
+import org.projectnessie.versioned.Operation;
+import org.projectnessie.versioned.ReferenceConflictException;
+import org.projectnessie.versioned.ReferenceNotFoundException;
+import org.projectnessie.versioned.VersionStore;
+import org.projectnessie.versioned.WithType;
+
+public class NamespaceApiImpl extends BaseApiImpl implements NamespaceApi {
+
+  public NamespaceApiImpl(
+      ServerConfig config,
+      VersionStore<Content, CommitMeta, Content.Type> store,
+      Authorizer authorizer,
+      Principal principal) {
+    super(config, store, authorizer, principal);
+  }
+
+  @Override
+  public Namespace createNamespace(NamespaceParams params)
+      throws NessieNamespaceAlreadyExistsException, NessieReferenceNotFoundException {
+    try {
+      BranchName branch = branchFromRefName(params);
+      if (getExplicitlyCreatedNamespace(params, branch).isPresent()) {
+        throw namespaceAlreadyExistsException(params);
+      }
+      if (getImplicitlyCreatedNamespace(params, branch).isPresent()) {
+        throw namespaceAlreadyExistsException(params);
+      }
+
+      Namespace namespace = params.getNamespace();
+      Put put = Put.of(ContentKey.of(namespace.getElements()), namespace);
+      commit(branch, "create namespace " + namespace.name(), TreeApiImpl.toOp(put));
+
+      return namespace;
+    } catch (ReferenceNotFoundException | ReferenceConflictException e) {
+      throw new NessieReferenceNotFoundException(e.getMessage(), e);
+    }
+  }
+
+  @Override
+  public void deleteNamespace(NamespaceParams params)
+      throws NessieReferenceNotFoundException, NessieNamespaceNotEmptyException,
```

Comment:
Similar here - "not empty check" is running outside the atomic commit operation.